posts/2022-08-04-jakarta-core-profile-22009-beta.adoc

```diff
@@ -254,10 +255,147 @@ More information about Expression Language 5.0 can be found in:
+import jakarta.inject.Inject;
+import jakarta.servlet.http.*;
+
+import java.sql.*;
+import java.util.concurrent.CompletionException;
+
+@ContextServiceDefinition(name = "java:app/concurrent/AppContextOnly",
+        propagated = APPLICATION,
+        cleared = { TRANSACTION, SECURITY },
+        unchanged = ALL_REMAINING)
+@ManagedExecutorDefinition(name = "java:module/concurrent/MyExecutor",
+        context = "java:app/concurrent/AppContextOnly",
+        maxAsync = 5)
+public class MyServlet extends HttpServlet {
+    @Resource(name = "java:module/env/concurrent/MyExecutorRef",
+              lookup = "java:module/concurrent/MyExecutor")
+    ManagedExecutorService myExecutor;
+
+    @Inject AccountsBean accounts;
+    ...
+----
+
+With the CDI feature enabled, methods that return a `CompletionStage` or `CompletableFuture` can be designated to run asynchronously by annotating them with `@Asynchronous`. For example:
+
+[source, java]
+----
+import jakarta.enterprise.concurrent.Asynchronous;
+import jakarta.inject.Singleton;
+
+import java.sql.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+import javax.sql.DataSource;
+
+@Singleton
+public class AccountsBean {
+
+    @Asynchronous(executor = "java:module/concurrent/MyExecutor")
+    CompletableFuture<List<Account>> findOverdue(int minDaysLate) {
+        List<Account> overdueAccounts = new ArrayList<Account>();
+        try {
+            // Resource reference lookup requires context of the application component,
+            DataSource db = InitialContext.doLookup("java:comp/env/jdbc/account-db-ref");
+            ...
+            return Asynchronous.Result.complete(overdueAccounts);
+        } catch (NamingException | SQLException x) {
+            throw new CompletionException(x);
+        }
+    }
+
+    ...
+----
+
+When the application invokes the asynchronous method, the container intercepts the invocation and hands back a completion stage representing its eventual completion, which is arranged by the `ManagedExecutorService`.  For example:
+
+[source, java]
+----
+    public void doGet(HttpServletRequest req, HttpServletResponse res) {
+        ...
+        accounts.findOverdue(60).thenAccept(this::sendReminder);
+        ...
+    }
+
+    void sendReminder(List<Account> overdue) {
+        try {
+            // Resource reference lookup requires context of the application component,
+            DataSource db = InitialContext.doLookup("java:comp/env/jdbc/customer-db-ref");
+            ...
+        } catch (NamingException | SQLException x) {
+            throw new CompletionException(x);
+        }
+    }
+----
+
+The Maven coordinates for Concurrency 3.0 are as follows:
+
+[source, xml]
+----
+<dependency>
+    <groupId>jakarta.enterprise.concurrent</groupId>
+    <artifactId>jakarta.enterprise.concurrent-api</artifactId>
+    <version>3.0.0</version>
```

Comment:
@njr-11 should this be updated to 3.0.1?

Reply:
There is no difference in the API between 3.0.0 and 3.0.1 and there will be a 3.0.2 created soon, but it doesn't exist yet.  Open Liberty is basically skipping over 3.0.1 and will go straight to 3.0.2 so I think it's fine to leave it.